plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/PyStringCodeCompletion.java

```diff
@@ -149,17 +156,42 @@
     /**
      * Needed interface for adding the completions on a request
      * @throws MisconfigurationException
+     * @throws PythonNatureWithoutProjectException
+     * @throws IOException
      */
     @Override
     public TokensOrProposalsList getCodeCompletionProposals(CompletionRequest request) throws CoreException,
-            BadLocationException, MisconfigurationException {
+            BadLocationException, MisconfigurationException, IOException, PythonNatureWithoutProjectException {
+
         List<ICompletionProposalHandle> completionProposals = new ArrayList<>();
         request.showTemplates = false; //don't show templates in strings
         fillWithEpydocFields(request, completionProposals);
 
         TokensOrProposalsList ret = new TokensOrProposalsList();
         if (completionProposals.size() == 0) {
             //if the size is not 0, it means that this is a place for the '@' stuff, and not for the 'default' context for a string.
+            IDocument doc = request.doc;
+            ITypedRegion partition = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc))
+                    .getPartition(request.documentOffset);
+            String partitionType = partition.getType();
+
+            if (partitionType.equals(IPythonPartitions.PY_MULTILINE_FSTRING1)
+                    || partitionType.equals(IPythonPartitions.PY_MULTILINE_FSTRING2)
+                    || partitionType.equals(IPythonPartitions.PY_SINGLELINE_FSTRING1)
+                    || partitionType.equals(IPythonPartitions.PY_SINGLELINE_FSTRING2)) {
+                int initOffset = request.documentOffset;
+                // To check whether where we are in the given completion offset,
+                // we are going to iterate backward from the initOffset
+
+                // if we find an initial bracket, it means that we are inside a fstring format or we are going to be inside a fstring format
+                for (int offset = initOffset; offset - 1 > partition.getOffset(); offset--) {
+                    char ch = doc.getChar(offset);
+                    if (ch == '{') {
```

Comment:
Humm, this doesn't seem very safe... if the user does something as:

`f'{b.bar()}  something else b.|<- cursor here'`

We don't want the completions (we just want the completions inside the parts of the f-string that can actually have code).

Also, we should ignore `{{`.

So, I think you should use the actual parser to parse this.

i.e.:
```
        FStringsGrammar fStringsGrammar = FStringsGrammarFactory.createGrammar(buf);
        FStringsAST ast = null;
        try {
            ast = fStringsGrammar.f_string();
        } catch (Throwable e) {
            // Just ignore any errors for this.
        }

        if (ast != null && ast.hasChildren()) {
              Iterable<SimpleNode> fStringExpressions = ast.getFStringExpressions();
              ... check if the cursor is in an f-string expression.
        }
```

You can see: `org.python.pydev.editor.PyFStringScanner.setRange(IDocument, int, int)` as a reference to a place which does something similar.